src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -575,5 +575,47 @@ private static IEnumerable<PropertyInfo> GetAllProperties(Type type)
 
             return allProperties;
         }
+
+        private static object GetPropertyValue(PropertyInfo property, object instance, IConfiguration config, BinderOptions options)
+        {
+            string propertyName = GetPropertyName(property);
+            return BindInstance(
+                property.PropertyType,
+                property.GetValue(instance),
+                config.GetSection(propertyName),
+                options);
+        }
+
+        private static string GetPropertyName(MemberInfo property)
+        {
+            if (property == null)
+            {
+                throw new ArgumentNullException(nameof(property));
+            }
+
+            foreach (var attributeData in property.GetCustomAttributesData())
+            {
+                if (attributeData.AttributeType != typeof(ConfigurationKeyNameAttribute) ||
+                    attributeData.ConstructorArguments.Count == 0)
+                {
+                    continue;
+                }
+
+                var constructorArgument = attributeData.ConstructorArguments.First();
+                if (constructorArgument.ArgumentType != typeof(string) ||
+                    constructorArgument.Value == null)
+                {
+                    continue;
```

Comment:
If we found a `ConfigurationKeyNameAttribute` with a null value, this would continue, meaning that we would iterate on all the attributes without needing to do so. 

I think we should break here and if the string is empty to just return the property name.

Reply:
I was wondering about that and if there could be a situation where there could be multiple key bindings whether it should just keep looping to the next one. If we don't want that behavior then adding the break makes total sense.

Reply:
I think we should just accept one key binding, and the attribute shouldn't accept multiple instances of it. So we should set allow multiple to false IMHO. It would be weird to have multiple key bindings and that the last found in metadata won.

Reply:
OK, that definitely solves the multiplicity problem. Will update here shortly.